### PR TITLE
[CIVP-11787] ENH futures.Executor classes for Civis

### PR DESCRIPTION
**Pull request by stephen-hoover**
Add two Civis-specific Executor classes (subclasses of the standard library `concurrent.futures.Executor`). The `_ContainerShellExecutor` uses Civis Platform container jobs to parallelize over either shell commands or container script arguments. The `CustomScriptExecutor` parallelizes over arguments to a custom script.

### CODE REVIEW

**Comment by stephen-hoover** [comment]
@keithing , this is a lot of lines, but I think only the second commit needs a line-by-line scrutiny. Also, I plan to add documentation on the new features to the markdown pages in a future PR.

File: civis/futures.py [hunk 1]
```diff
@@ -285,3 +292,386 @@ def cancel(self):
                 self._invoke_callbacks()
                 return self.cancelled()
             return False
+
+
+def create_docker_command(*args, **kwargs):
```
Comment (keithing) [inline civis/futures.py]: Do we want this to be private as well?

Comment (stephen-hoover) [inline civis/futures.py]: You're right, it should be. Fixed.

File: civis/futures.py [hunk 2]
```diff
@@ -285,3 +292,386 @@ def cancel(self):
...
+            hidden=self.hidden,
+            params=self.params,
+            arguments=combined_args,
+            git_credential_id=self.git_credential_id,
+        )
+
+        return job
+
+
+class CustomPoolExecutor(_CivisExecutor):
```
Comment (keithing) [inline civis/futures.py]: `CustomScriptPoolExecutor`?

File: civis/futures.py [hunk 3]
```diff
@@ -285,3 +292,386 @@ def cancel(self):
...
+                 max_n_retries=0,
+                 client=None,
+                 polling_interval=None,
+                 inc_script_names=False):
+        self.from_template_id = from_template_id
+        self.arguments = arguments
+
+        if script_name is None:
+            date_str = datetime.datetime.today().strftime("%Y-%m-%d")
+            script_name = "CustomPoolExecutorScript {}".format(date_str)
```
Comment (keithing) [inline civis/futures.py]: If we change the name, we need to update that here too.

**Comment by keithing** [comment]
Everything looks good, after we decide what the correct name is for `CustomPoolExecutor` this looks ready to go.

File: civis/futures.py (hunk 2 above)
Comment (stephen-hoover) [inline civis/futures.py]: How about `CustomScriptExecutor`? It's shorter, and there's not really a "pool".

Comment (keithing) [inline civis/futures.py]: Great, I like that.

**Comment by stephen-hoover** [comment]
I renamed `CustomPoolExecutor` to `CustomScriptExecutor`. I also changed `_ContainerPoolShellExecutor` to `_ContainerShellExecutor`; it's shorter and probably more accurate.

File: civis/futures.py [hunk 4]
```diff
@@ -285,3 +292,386 @@ def cancel(self):
...
+        self.params = params
+        self.arguments = arguments
+
+        if required_resources is None:
+            required_resources = {'cpu': 1024, 'memory': 1024}
+        self.required_resources = required_resources
+
+        if script_name is None:
+            date_str = datetime.datetime.today().strftime("%Y-%m-%d")
+            script_name = ("ContainerPoolShellExecutorScript "
```
Comment (keithing) [inline civis/futures.py]: This needs to be renamed too.

Comment (stephen-hoover) [inline civis/futures.py]: Oops. Fixed.

**Pull request merged**